File: src/lib/builds/equipPlan.test.ts
import { describe, expect, it, vi } from "vitest";

import { planEquipSteps } from "@/lib/builds/equipPlan";
import { executeEquipPlan } from "@/lib/builds/equipOrchestrator";
import { createMockWriteClient } from "@/lib/bungie/writeClient";
import type { UserInventoryItem } from "@/lib/db/types";

function inv(
  partial: Pick<UserInventoryItem, "instanceId" | "itemHash" | "location"> & {
    characterId?: string;
  },
): UserInventoryItem {
  return {
    bucket: "Helmet",
    power: 100,
    isMasterwork: false,
    isCrafted: false,
    plugHashes: [],
    rollTags: [],
    syncedAt: "2026-01-01T00:00:00.000Z",
    ...partial,
  };
}

describe("planEquipSteps (US2)", () => {
  it("orders transfer → equip → artifact → fashion", () => {
    const plan = planEquipSteps({
      characterId: "char-a",
      equipment: {
        helmet: {
          slot: "helmet",
          itemHash: 10,
          itemName: "Helm",
          source: "set",
          instanceId: "i-helm",
        },
      },
      artifact: { hash: 99, name: "Artifact", config: [1, 2] },
      fashion: {
        setId: "f1",
        slots: { ghost: { itemHash: 50, itemName: "Ghost" } },
      },
      inventory: [
        inv({ instanceId: "i-helm", itemHash: 10, location: "vault" }),
        inv({ instanceId: "i-ghost", itemHash: 50, location: "vault" }),
      ],
    });

    expect(plan.map((s) => s.kind)).toEqual(["transfer", "equip", "artifact", "fashion"]);
    expect(plan[0]?.id).toBe("transfer-helmet");
    expect(plan[0]?.transferToVault).toBe(false);
    expect(plan[1]?.id).toBe("equip-helmet");
    expect(plan[2]?.id).toBe("artifact");
    expect(plan[3]?.id).toBe("fashion-ghost");
  });

  it("vault-hops when item is on another character", () => {
    const plan = planEquipSteps({
      characterId: "char-a",
      equipment: {
        primary: {
          slot: "primary",
          itemHash: 1,
          itemName: "Gun",
          source: "set",
          instanceId: "i-gun",
        },
      },
      artifact: null,
      fashion: null,
      inventory: [
        inv({
          instanceId: "i-gun",
          itemHash: 1,
          location: "equipped",
          characterId: "char-b",
        }),
      ],
    });

    expect(plan.map((s) => s.id)).toEqual([
      "transfer-primary-to-vault",
      "transfer-primary-from-vault",
      "equip-primary",
    ]);
  });

  it("skips transfer when already on target character", () => {
    const plan = planEquipSteps({
      characterId: "char-a",
      equipment: {
        arms: {
          slot: "arms",
          itemHash: 2,
          itemName: "Arms",
          source: "set",
          instanceId: "i-arms",
        },
      },
      artifact: null,
      fashion: null,
      inventory: [
        inv({
          instanceId: "i-arms",
          itemHash: 2,
          location: "character",
          characterId: "char-a",
        }),
      ],
    });

    expect(plan.map((s) => s.kind)).toEqual(["equip"]);
  });

  it("omits empty fashion slots and missing artifact", () => {
    const plan = planEquipSteps({
      characterId: "char-a",
      equipment: {},
      artifact: null,
      fashion: { setId: "f1", slots: {} },
      inventory: [],
    });
    expect(plan).toEqual([]);
  });
});

describe("executeEquipPlan (US2/US3)", () => {
  const ctx = { accessToken: "t", membershipType: 3 };

  it("executes steps via write client in plan order", async () => {
    const calls: string[] = [];
    const client = createMockWriteClient({
      transferItem: async () => {
        calls.push("transfer");
      },
      equipItem: async () => {
        calls.push("equip");
      },
      applyArtifactConfig: async () => {
        calls.push("artifact");
      },
      applyFashionSlot: async () => {
        calls.push("fashion");
      },
    });

    const status = await executeEquipPlan(client, ctx, "char-a", [
      {
        id: "transfer-helmet",
        kind: "transfer",
        itemHash: 10,
        instanceId: "i1",
        transferToVault: false,
      },
      { id: "equip-helmet", kind: "equip", itemHash: 10, instanceId: "i1" },
      { id: "artifact", kind: "artifact", itemHash: 99, artifactConfig: [1] },
      { id: "fashion-ghost", kind: "fashion", slot: "ghost", itemHash: 50 },
    ]);

    expect(calls).toEqual(["transfer", "equip", "artifact", "fashion"]);
    expect(status.completed).toBe(4);
    expect(status.failed).toBe(0);
  });

  it("keeps prior ok steps when a later step fails (partial status)", async () => {
    const client = createMockWriteClient({
      equipItem: async (_ctx, args) => {
        if (args.instanceId === "fail") throw new Error("bungie denied");
      },
    });

    const status = await executeEquipPlan(client, ctx, "char-a", [
      {
        id: "transfer-helmet",
        kind: "transfer",
        itemHash: 10,
        instanceId: "ok",
        transferToVault: false,
      },
      { id: "equip-helmet", kind: "equip", itemHash: 10, instanceId: "fail" },
      { id: "equip-arms", kind: "equip", itemHash: 11, instanceId: "ok2" },
    ]);

    expect(status.steps[0]?.ok).toBe(true);
    expect(status.steps[1]?.ok).toBe(false);
    expect(status.steps[1]?.error).toMatch(/bungie denied/);
    expect(status.steps[2]?.ok).toBe(true);
    expect(status.completed).toBe(2);
    expect(status.failed).toBe(1);
  });
});

void vi;

File: packages/domain/test/equip_plan_test.dart
import 'package:destiny2_domain/destiny2_domain.dart';
import 'package:test/test.dart';

SlotClaim claim({
  required EquipmentSlot slot,
  required int itemHash,
  required String instanceId,
  String itemName = 'Item',
}) {
  return SlotClaim(
    slot: slot,
    itemHash: itemHash,
    itemName: itemName,
    source: ClaimSource.set,
    instanceId: instanceId,
  );
}

EquipInventoryItem inv({
  required String instanceId,
  required int itemHash,
  required String location,
  String? characterId,
}) {
  return EquipInventoryItem(
    instanceId: instanceId,
    itemHash: itemHash,
    location: location,
    characterId: characterId,
  );
}

void main() {
  group('planEquipSteps (US1)', () {
    test('orders transfer → equip → artifact → fashion', () {
      final plan = planEquipSteps(
        EquipPlanInput(
          characterId: 'char-a',
          equipment: {
            EquipmentSlot.helmet: claim(
              slot: EquipmentSlot.helmet,
              itemHash: 10,
              instanceId: 'i-helm',
              itemName: 'Helm',
            ),
          },
          artifact: const EquipPlanArtifact(
            hash: 99,
            name: 'Artifact',
            config: [1, 2],
          ),
          fashion: const EquipPlanFashion(
            setId: 'f1',
            slots: {
              'ghost': EquipPlanFashionPiece(itemHash: 50, itemName: 'Ghost'),
            },
          ),
          inventory: [
            inv(instanceId: 'i-helm', itemHash: 10, location: 'vault'),
            inv(instanceId: 'i-ghost', itemHash: 50, location: 'vault'),
          ],
        ),
      );

      expect(
        plan.map((s) => s.kind).toList(),
        [
          EquipStepKind.transfer,
          EquipStepKind.equip,
          EquipStepKind.artifact,
          EquipStepKind.fashion,
        ],
      );
      expect(plan[0].id, 'transfer-helmet');
      expect(plan[0].transferToVault, isFalse);
      expect(plan[1].id, 'equip-helmet');
      expect(plan[2].id, 'artifact');
      expect(plan[3].id, 'fashion-ghost');
      expect(plan[3].instanceId, 'i-ghost');
    });

    test('vault-hops when item is on another character', () {
      final plan = planEquipSteps(
        EquipPlanInput(
          characterId: 'char-a',
          equipment: {
            EquipmentSlot.primary: claim(
              slot: EquipmentSlot.primary,
              itemHash: 1,
              instanceId: 'i-gun',
              itemName: 'Gun',
            ),
          },
          inventory: [
            inv(
              instanceId: 'i-gun',
              itemHash: 1,
              location: 'equipped',
              characterId: 'char-b',
            ),
          ],
        ),
      );

      expect(
        plan.map((s) => s.id).toList(),
        [
          'transfer-primary-to-vault',
          'transfer-primary-from-vault',
          'equip-primary',
        ],
      );
      expect(plan[0].transferToVault, isTrue);
      expect(plan[1].transferToVault, isFalse);
    });

    test('skips transfer when already on target character', () {
      final plan = planEquipSteps(
        EquipPlanInput(
          characterId: 'char-a',
          equipment: {
            EquipmentSlot.arms: claim(
              slot: EquipmentSlot.arms,
              itemHash: 2,
              instanceId: 'i-arms',
              itemName: 'Arms',
            ),
          },
          inventory: [
            inv(
              instanceId: 'i-arms',
              itemHash: 2,
              location: 'character',
              characterId: 'char-a',
            ),
          ],
        ),
      );

      expect(plan.map((s) => s.kind).toList(), [EquipStepKind.equip]);
    });

    test('omits empty fashion slots and missing artifact', () {
      final plan = planEquipSteps(
        const EquipPlanInput(
          characterId: 'char-a',
          equipment: {},
          fashion: EquipPlanFashion(setId: 'f1'),
          inventory: [],
        ),
      );
      expect(plan, isEmpty);
    });

    test(
      'throws NOT_EQUIP_READY when combat instanceId is missing from inventory',
      () {
        expect(
          () => planEquipSteps(
            EquipPlanInput(
              characterId: 'char-a',
              equipment: {
                EquipmentSlot.helmet: claim(
                  slot: EquipmentSlot.helmet,
                  itemHash: 10,
                  instanceId: 'missing-helm',
                  itemName: 'Helm',
                ),
              },
              inventory: const [],
            ),
          ),
          throwsA(
            isA<EquipReadyException>().having(
              (e) => e.code,
              'code',
              DomainFailureCodes.notEquipReady,
            ),
          ),
        );
      },
    );

    test('plans equip steps for all pinned combat slots on happy path', () {
      final plan = planEquipSteps(
        EquipPlanInput(
          characterId: 'char-a',
          equipment: {
            EquipmentSlot.primary: claim(
              slot: EquipmentSlot.primary,
              itemHash: 1,
              instanceId: 'i-primary',
            ),
            EquipmentSlot.helmet: claim(
              slot: EquipmentSlot.helmet,
              itemHash: 10,
              instanceId: 'i-helm',
            ),
            EquipmentSlot.arms: claim(
              slot: EquipmentSlot.arms,
              itemHash: 11,
              instanceId: 'i-arms',
            ),
          },
          inventory: [
            inv(
              instanceId: 'i-primary',
              itemHash: 1,
              location: 'character',
              characterId: 'char-a',
            ),
            inv(
              instanceId: 'i-helm',
              itemHash: 10,
              location: 'character',
              characterId: 'char-a',
            ),
            inv(
              instanceId: 'i-arms',
              itemHash: 11,
              location: 'character',
              characterId: 'char-a',
            ),
          ],
        ),
      );

      expect(
        plan
            .where((s) => s.kind == EquipStepKind.equip)
            .map((s) => s.id)
            .toList(),
        ['equip-primary', 'equip-helmet', 'equip-arms'],
      );
    });

    test('skips combat claims without instanceId (wishlist)', () {
      final plan = planEquipSteps(
        EquipPlanInput(
          characterId: 'char-a',
          equipment: {
            EquipmentSlot.primary: const SlotClaim(
              slot: EquipmentSlot.primary,
              itemHash: 1,
              itemName: 'Gun',
              source: ClaimSource.set,
            ),
          },
          inventory: const [],
        ),
      );
      expect(plan, isEmpty);
    });
  });
}
